### PR TITLE
DDF-3025 Upgrade to Pax Exam 4.11.0

### PR DESCRIPTION
**Pull request by vinamartin**
#### What does this PR do?
Pax Exam 4.11.0 Release notes: https://ops4j1.jira.com/wiki/display/PAXEXAM4/2017/05/16/Pax+Exam+4.11.0+released

Changes by file: 

_catalog-app/src/main/resources/features.xml_
Rearranges bundles in catalog-core to prevent a timing issue

_distribution/test/itests/pom.xml_
Updates the pax exam version to 4.11.0, added a comment

_test-itests-common/src/main/java/org/codice/ddf/itests/common/AbstractIntegrationTest.java_
Fixes an issue with a configuration that gets set using pax exam that used to be in quotations but no longer needs to be.

_test-itests-common/src/main/java/org/codice/ddf/itests/common/ServiceManagerProxy.java_
Creates a retry for getting the SecurityManager in the ServiceManagerProxy to prevent a potential timing issue that was occurring in the itests. This issue currently exists in master, but was happening more frequently with the upgrade. 

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
@emmberk @kcwire @lessarderic 

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@clockard
@stustison

#### How should this be tested? (List steps with links to updated documentation)
Run the integration tests and make sure nothing fails

#### What are the relevant tickets?
[DDF-3025](https://codice.atlassian.net/browse/DDF-3025)

### CODE REVIEW

File: catalog/core/catalog-core-camelcomponent/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 1]
```diff
@@ -17,11 +17,11 @@
   http://www.osgi.org/xmlns/blueprint/v1.0.0 http://www.osgi.org/xmlns/blueprint/v1.0.0/blueprint.xsd"
         >
 
-    <reference id="transformerMapper" interface="ddf.mime.MimeTypeToTransformerMapper"/>
+    <reference id="transformerMapper" interface="ddf.mime.MimeTypeToTransformerMapper" availability="optional"/>
```
Comment (emmberk) [inline catalog/core/catalog-core-camelcomponent/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: @vinamartin, I think the only one in this file that needs `optional` added is the CatalogFramework.

File: distribution/test/itests/pom.xml [hunk 2]
```diff
@@ -31,9 +31,12 @@
         <module>test-itests-dependencies-app</module>
     </modules>
     <properties>
-        <pax.exam.version>4.9.2</pax.exam.version>
+        <!-- Copied from ops4j/org.ops4j.pax.exam2/pom/pom.xml
```
Comment (emmberk) [inline distribution/test/itests/pom.xml]: We should include a version number in this comment, like we did [here](https://github.com/codice/ddf/blob/master/platform/platform-app/src/main/resources/features.xml#L36).

File: catalog/core/catalog-core-camelcomponent/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 1 above)
Comment (vinamartin) [inline catalog/core/catalog-core-camelcomponent/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Sounds good

File: distribution/test/itests/pom.xml (hunk 2 above)
Comment (vinamartin) [inline distribution/test/itests/pom.xml]: Will update

**Comment by vinamartin** [comment]
build now

File: catalog/core/catalog-core-camelcomponent/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 3]
```diff
@@ -15,32 +15,34 @@
...
   http://www.osgi.org/xmlns/blueprint/v1.0.0 http://www.osgi.org/xmlns/blueprint/v1.0.0/blueprint.xsd"
-        >
+>
 
     <reference id="transformerMapper" interface="ddf.mime.MimeTypeToTransformerMapper"/>
 
     <reference id="mimeTypeMapper" interface="ddf.mime.MimeTypeMapper"/>
 
-    <reference id="framework" interface="ddf.catalog.CatalogFramework"/>
+    <reference id="framework" interface="ddf.catalog.CatalogFramework" availability="optional"/>
```
Comment (kcwire) [inline catalog/core/catalog-core-camelcomponent/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: My concern is if the code that is using the `framework` reference is protecting for when it is not available. If the `catalogComponent` bean were to be called before the framework is available will it handle that gracefully. Same comment for the other references that were changed to `optional`.

Comment (vinamartin) [inline catalog/core/catalog-core-camelcomponent/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: I agree, I was pretty hesitant about making this change. What's the best way to handle this gracefully? Null checks?

Comment (kcwire) [inline catalog/core/catalog-core-camelcomponent/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: https://stackoverflow.com/questions/10794082/blueprint-how-do-i-check-if-an-optional-dependency-is-satisfied

Comment (vinamartin) [inline catalog/core/catalog-core-camelcomponent/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: There seems to be some kind of issue with how the bundles in [catalog-core](https://github.com/codice/ddf/blob/master/catalog/catalog-app/src/main/resources/features.xml#L41) are being started, which seems unique to the [itests.](https://github.com/codice/ddf/blob/master/distribution/test/itests/test-itests-dependencies-app/src/main/filtered-resources/features.xml#L27) The problem is that bundles are waiting for other bundles that are started in that feature to become available, and it doesn't look like we are able to impose an order on them. I've replicating what the itests do outside of pax-exam and everything resolves fine, which makes me think there is a timing issue here.

Comment (emmberk) [inline catalog/core/catalog-core-camelcomponent/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: @vinamartin, it might be worth adding a unit test to check the behavior when the `catalogFramework` is null.

Comment (emmberk) [inline catalog/core/catalog-core-camelcomponent/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: @vinamartin, we could also add a unit test for this case.

File: catalog/catalog-app/src/main/resources/features.xml [hunk 4]
```diff
@@ -79,6 +71,19 @@
         <bundle>mvn:ddf.catalog.core/versioning-api/${project.version}</bundle>
     </feature>
 
+    <feature name="catalog-core-plugins" install="auto" version="${project.version}"
```
Comment (vinamartin) [inline catalog/catalog-app/src/main/resources/features.xml]: Looking for a better name and description suggestion here

**Comment by vinamartin** [comment]
build now

**Comment by vinamartin** [comment]
build now

File: platform/security/common/src/main/java/org/codice/ddf/security/common/Security.java [hunk 5]
```diff
@@ -299,12 +299,28 @@ public Date getExpires(Subject subject) {
...
                     context.getServiceReference(SecurityManager.class);
-            return (SecurityManager) context.getService(securityManagerRef);
+
+            int tries = 0;
+            while (tries < 10) {
+                try {
+                    SecurityManager securityManager = (SecurityManager) context.getService(
+                            securityManagerRef);
+                    return securityManager;
+                } catch (NullPointerException e) {
```
Comment (lessarderic) [inline platform/security/common/src/main/java/org/codice/ddf/security/common/Security.java]: ❓ Anything we can check to prevent the `NullPointerException`? Catching it seems a little odd and may not work in the future.

File: platform/security/common/src/main/java/org/codice/ddf/security/common/Security.java [hunk 6]
```diff
@@ -299,12 +299,28 @@ public Date getExpires(Subject subject) {
      */
     public SecurityManager getSecurityManager() {
         BundleContext context = getBundleContext();
+
         if (context != null) {
             ServiceReference securityManagerRef =
                     context.getServiceReference(SecurityManager.class);
-            return (SecurityManager) context.getService(securityManagerRef);
+
+            int tries = 0;
```
Comment (lessarderic) [inline platform/security/common/src/main/java/org/codice/ddf/security/common/Security.java]: ❗️ We should use https://github.com/jhalterman/failsafe instead of hand-coding our own retry logic.

File: catalog/catalog-app/src/main/resources/features.xml (hunk 4 above)
Comment (clockard) [inline catalog/catalog-app/src/main/resources/features.xml]: I think catalog-core-plugins is fine

File: catalog/core/catalog-core-camelcomponent/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 7]
```diff
@@ -15,32 +15,34 @@
            xmlns="http://www.osgi.org/xmlns/blueprint/v1.0.0"
            xsi:schemaLocation="
   http://www.osgi.org/xmlns/blueprint/v1.0.0 http://www.osgi.org/xmlns/blueprint/v1.0.0/blueprint.xsd"
-        >
+>
```
Comment (clockard) [inline catalog/core/catalog-core-camelcomponent/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Was formatting the only change to this file?

File: distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/ServiceManagerProxy.java [hunk 8]
```diff
@@ -38,8 +42,15 @@ public ServiceManagerProxy(ServiceManager serviceManager) {
 
     @Override
     public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
-
-        Subject subject = SECURITY.runAsAdmin(SECURITY::getSystemSubject);
-        return subject.execute(() -> method.invoke(serviceManager, args));
+        final RetryPolicy retryPolicy = new RetryPolicy().retryOn(NullPointerException.class)
+                .withDelay(1, TimeUnit.SECONDS)
+                .withMaxRetries(10);
```
Comment (clockard) [inline distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/ServiceManagerProxy.java]: Wonder if we might need a higher number for very slow systems. Probably ok as is.

File: catalog/core/catalog-core-camelcomponent/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 7 above)
Comment (vinamartin) [inline catalog/core/catalog-core-camelcomponent/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Was just about to remove it

File: catalog/core/catalog-core-camelcomponent/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 9]
```diff
@@ -26,7 +26,7 @@
     <reference id="uuidGenerator" interface="org.codice.ddf.platform.util.uuidgenerator.UuidGenerator" filter="(id=uuidGenerator)"/>
 
     <!--
-    Instantiate our custom Camel CatalogComponent and inject this bundle context into it.  
+    Instantiate our custom Camel CatalogComponent and inject this bundle context into it.
```
Comment (vinamartin) [inline catalog/core/catalog-core-camelcomponent/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: I can't seem to get rid of this space. Not part of the change set.

**Comment by vinamartin** [comment]
@lessarderic could you give this another look?

**Comment by vinamartin** [comment]
build now

File: distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/ServiceManagerProxy.java [hunk 10]
```diff
@@ -38,8 +42,15 @@ public ServiceManagerProxy(ServiceManager serviceManager) {
 
     @Override
     public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
-
-        Subject subject = SECURITY.runAsAdmin(SECURITY::getSystemSubject);
-        return subject.execute(() -> method.invoke(serviceManager, args));
+        final RetryPolicy retryPolicy = new RetryPolicy().retryOn(NullPointerException.class)
```
Comment (lessarderic) [inline distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/ServiceManagerProxy.java]: ❗️ Since we've moved this retry logic in the tests instead of the code, we should replace failsafe with https://github.com/awaitility/awaitility. Sorry :-(

Comment (emmberk) [inline distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/ServiceManagerProxy.java]: @clockard added this retry in a separate PR now that this error is occurring in itests even without the pax exam changes. I have rebased this PR from https://github.com/codice/ddf/pull/2015.

**Comment by vinamartin** [comment]
build now

**Comment by vinamartin** [comment]
build now
